conda_smithy/configure_feedstock.py

```diff
@@ -609,22 +618,27 @@ def _collapse_subpackage_variants(
     all_used_vars.update(always_keep_keys)
     all_used_vars.update(top_level_vars)
 
-    logger.debug(
-        "final all_used_vars {}".format(pprint.pformat(all_used_vars))
-    )
-    logger.debug("top_level_vars {}".format(pprint.pformat(top_level_vars)))
-    logger.debug(
-        "top_level_loop_vars {}".format(pprint.pformat(top_level_loop_vars))
-    )
+    _log_debug_with_pprint("final all_used_vars:", all_used_vars)
 
-    used_key_values = _get_used_key_values_by_input_order(
+    used_key_values, used_zipped_vars = _get_used_key_values_by_input_order(
         squished_input_variants,
         squished_used_variants,
         all_used_vars,
     )
 
     for k in preserve_top_level_loops:
-        used_key_values[k] = squished_input_variants[k]
+        # so sometimes a key in a zip will not be used in the recipe and so
+        # gets put into preserve_top_level_loops.
+        # however, if we assign it to used_key_values, the careful reordering
+        # of the zipping in `_get_used_key_values_by_input_order` gets undone.
+        # so we skip it here
+
+        # TODO - this is a hack and needs fixing
+        # it turns out sometimes conda-build just doesn't return some of the variants
+        # even if they are in the zipped vars. So we also keep any key which on output
+        # has a length of 1 but had length > 1 on input.
```

Comment:
@mbargull Can you look the issue here over? 

In particular, IDK why conda-build when it computes all of the outputs making a list of metas, sometimes does and sometimes does not return outputs for the all of the key values in preserve_top_level_loops. 

Right now I put in a hack to get the tests to pass but I suspect there is something more fundamental going on that needs fixing.

Reply:
I figure this out I think. When conda-build renders a recipe, it returns the metadata for each unique file generated. If a key we use is zipped with a key we do not use, conda-build will ignore the variants involving all but one of the values of the key we do not use. This behavior is not what we do in conda-forge (i.e., we want all variants that are not explicitly skipped even if some of the keys in the variants are not explicitly used).

The most robust way to fix this is to write a custom function that returns metadata for each of the variants in the full exploded matrix except the ones that the recipe skips.

Reply:
The part about zip keys is wrong above. It is a similar issue but about top-level keys not used by outputs.